WIP [IOTEVENT-953] jwt token validator

tylershaosamsung opened the pull request.

tylershaosamsung committed:
[IOTEVENT-953] jwt token validator

beckje01 commented on ratpack-bearer-auth/src/main/java/st/ratpack/auth/springsec/jwt/JWTTokenValidator.java:
```diff
@@ -0,0 +1,72 @@
+
+	@Override
+	public Promise<Optional<OAuthToken>> validate(String token) {
+		if (isJwtToken(token) && !verifyJwtToken(token)) {
```
It's a good start I think it may make sense to allow for a predicate and separate JWT validation from the I'll spike something out

beckje01 commented on ratpack-bearer-auth/src/main/java/st/ratpack/auth/springsec/SpringSecCheckAuthModule.java:
```diff
@@ -3,8 +3,10 @@
 import st.ratpack.auth.*;
+import st.ratpack.auth.springsec.jwt.DefaultJWKProvider;
+import st.ratpack.auth.springsec.jwt.JWKProvider;
+import st.ratpack.auth.springsec.jwt.JWTTokenValidator;
```
JWT isn't very common with SpringSecurity we probably don't want to have this in the spring security module.

tylershaosamsung commented:
@beckje01 for the comment about not having jwt in spring security module. i changed to have a seperate jwt module, which has a interceptor cuts the invocation of TokenProvider.checkToken to verify the jwt token string. This is with assumption that we may have other type of auth modules other than spring security in the future, so i didnt make a new validator in jwt module which just inherits SpringSecCheckTokenValidator. 

@Bijnagte @jeff-blaisdell @edwardsoo-ss 

I am still working on other comment about separation JWT validation to somewhere else

beckje01 commented:
I was thinking something like this to help with composition of validators https://github.com/SmartThingsOSS/ratpack-bearer-auth/pull/37

tylershaosamsung commented:
@beckje01 @Bijnagte 
added an update (WIP) to make JOSE jwt verification code more flexible, ex, support different jwt token types (JWS, JWE, UnsecuredJWT). also moved ratpack jwt module to [the MR in auth project](https://git.vandevlab.com/iot/auth/merge_requests/54/diffs#8af3d7721e4374783eec49125d264c3496c14318_0_25) which wraps jwt verfication code implemented in this repo with spring sec validator using OrderedPredicateTokenValidator. and OSS developers can make their own modules with similiar approach.

the code is not finalized yet, but please comment any suggestions

one thing need to discuss is if its good idea to put core jwt verification code in this repo or not, according to @llinder 's early comment, that most of the JWT validation should be non framework specific so we can more easily adapt it to other frameworks. However i can not extract common logics to auth repo which is internal thus not accessible from open source library

tylershaosamsung commented on ratpack-bearer-auth/src/main/java/st/ratpack/auth/PredicateValidatorTuple.java:
```diff
@@ -0,0 +1,28 @@
+import java.util.function.Predicate;
+
+public class PredicateValidatorTuple {
+    private Predicate<String> predicate;
```
temporarily copied @beckje01 's PredicateValidatorTuple for testing purpose, until https://github.com/SmartThingsOSS/ratpack-bearer-auth/pull/37 is merged

tylershaosamsung commented:
actually putting jwt validation logics in this repo seems not good, according to @llinder 's comment on https://git.vandevlab.com/iot/auth/merge_requests/54/diffs#8af3d7721e4374783eec49125d264c3496c14318_0_25, most of the JWT validation should be non framework specific so we can more easily adapt it to other frameworks. should i move jwt validation logics to auth repo and not open source it for now?

tylershaosamsung commented:
closing the PR. as discussed with @Bijnagte , will move the code to auth repository for now
